src/vcs/git/repo.rs

```diff
@@ -0,0 +1,521 @@
+// This file is part of radicle-surf
+// <https://github.com/radicle-dev/radicle-surf>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+pub use crate::vcs::git::object::{Branch, BranchName, Commit, RevObject, Signature, TagName};
+use crate::{
+    diff,
+    diff::*,
+    file_system,
+    tree::*,
+    vcs,
+    vcs::{git::error::*, VCS},
+};
+use git2::{BranchType, Oid};
+use nonempty::NonEmpty;
+use std::{cmp::Ordering, convert::TryFrom, ops::Deref, str};
+
+/// OrderedCommit is to allow for us to identify an ordering of commit history
+/// as we enumerate over a revwalk of commits, by assigning each commit an
+/// identifier.
+#[derive(Clone)]
+pub(super) struct OrderedCommit {
+    pub(super) id: usize,
+    pub(super) commit: Commit,
+}
+
+impl std::fmt::Debug for OrderedCommit {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "OrderedCommit {{ id: {}, commit: {} }}",
+            self.id, self.commit.id
+        )
+    }
+}
+
+impl OrderedCommit {
+    pub(super) fn compare_by_id(&self, other: &Self) -> Ordering {
+        self.id.cmp(&other.id).reverse()
+    }
+}
+
+impl From<OrderedCommit> for Commit {
+    fn from(ordered_commit: OrderedCommit) -> Self {
+        ordered_commit.commit
+    }
+}
+
+/// A `History` that uses `git2::Commit` as the underlying artifact.
+pub type History = vcs::History<Commit>;
+
+/// Wrapper around the `git2`'s `git2::Repository` type.
+/// This is to to limit the functionality that we can do
+/// on the underlying object.
+pub struct Repository(pub(super) git2::Repository);
+
+/// A reference-only `Repository`. This means that we cannot mutate the
+/// underlying `Repository`. Not being able to mutate the `Repository` means
+/// that the functions defined for `RepositoryRef` should be thread-safe.
+pub struct RepositoryRef<'a> {
+    pub(super) repo_ref: &'a git2::Repository,
+}
+
+impl<'a> Deref for RepositoryRef<'a> {
```

Comment:
Ah wait, I think this is a bit dangerous as it could break the readonly contract

Reply:
But it returns a reference so it's still read-only no?

Reply:
Ah no, you're right. There's `&self` functions that mutate the underlying repository data

Reply:
Dropped the commit